src/box/iproto.cc

```diff
@@ -706,7 +808,16 @@ iproto_connection_close(struct iproto_connection *con)
 		 * is done only once.
 		 */
 		con->p_ibuf->wpos -= con->parse_size;
-		cpipe_push(&con->iproto_thread->tx_pipe, &con->disconnect_msg);
+		/**
+		 * If there is no streams we immediately push
+		 * disconnect_msg to tx thread, otherwise it will be
+		 * pushed from `net_finish_destroy_stream_on_disconnect`
+		 * when last stream will be destroyed.
+		 */
+		if (mh_size(con->streams) == 0) {
+			cpipe_push(&con->iproto_thread->tx_pipe,
```

Comment:
We need iproto_session `tx_process_destroy_stream_on_disconnect`, but it can be destroyed if we process con->disconnect_msg before stream disconnect message.

Reply:
After private discussion i fix this!